aptos-move/aptos-vm/src/data_cache.rs

```diff
@@ -65,32 +76,40 @@ impl<'a, S: StateView> StorageAdapter<'a, S> {
         address: &AccountAddress,
         struct_tag: &StructTag,
         metadata: &[Metadata],
-    ) -> Result<Option<Vec<u8>>, VMError> {
+    ) -> Result<(Option<Vec<u8>>, Option<NumBytes>), VMError> {
         let resource_group = get_resource_group_from_metadata(struct_tag, metadata);
         if let Some(resource_group) = resource_group {
             let mut cache = self.resource_group_cache.borrow_mut();
             let cache = cache.entry(*address).or_insert_with(BTreeMap::new);
             if let Some(group_data) = cache.get_mut(&resource_group) {
                 // This resource group is already cached for this address. So just return the
                 // cached value.
-                return Ok(group_data.get(struct_tag).cloned());
+                let buf = group_data.get(struct_tag).cloned();
+                let len = buf.as_ref().map(|b| NumBytes::new(b.len() as u64));
+                return Ok((buf, len));
             }
             let group_data = self.get_resource_group_data(address, &resource_group)?;
             if let Some(group_data) = group_data {
+                let mut len = Some(NumBytes::new(group_data.len() as u64));
                 let group_data: BTreeMap<StructTag, Vec<u8>> = bcs::from_bytes(&group_data)
                     .map_err(|_| {
                         PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
                             .finish(Location::Undefined)
                     })?;
                 let res = group_data.get(struct_tag).cloned();
+                if !self.accurate_byte_count.get() {
+                    len = res.as_ref().map(|b| NumBytes::new(b.len() as u64));
+                }
```

Comment:
Yeah, as we discussed offline, let's try to make the amount charged independent of the ordering by which you load the resources in the same group.